`lib/transform` and `convmv`

nielash opened the pull request:
#### What is the purpose of this change?

Adds the `lib/transform` library and `convmv` command for converting and renaming files and directories by applying prefixes, suffixes, and other alterations.

#### Was the change discussed in an issue or in the forum before?

- https://github.com/rclone/rclone/issues/8393#issuecomment-2659988613
- https://github.com/rclone/rclone/issues/7072#issuecomment-1941707626

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)


## Synopsis


convmv supports advanced path name transformations for converting and renaming files and directories by applying prefixes, suffixes, and other alterations.

| Command | Description |
|------|------|
| `--name-transform prefix=XXXX` | Prepends XXXX to the file name. |
| `--name-transform suffix=XXXX` | Appends XXXX to the file name after the extension. |
| `--name-transform suffix_keep_extension=XXXX` | Appends XXXX to the file name while preserving the original file extension. |
| `--name-transform trimprefix=XXXX` | Removes XXXX if it appears at the start of the file name. |
| `--name-transform trimsuffix=XXXX` | Removes XXXX if it appears at the end of the file name. |
| `--name-transform regex=/pattern/replacement/` | Applies a regex-based transformation. |
| `--name-transform replace=old:new` | Replaces occurrences of old with new in the file name. |
| `--name-transform date={YYYYMMDD}` | Appends or prefixes the specified date format. |
| `--name-transform truncate=N` | Truncates the file name to a maximum of N characters. |
| `--name-transform base64encode` | Encodes the file name in Base64. |
| `--name-transform base64decode` | Decodes a Base64-encoded file name. |
| `--name-transform encoder=ENCODING` | Converts the file name to the specified encoding (e.g., ISO-8859-1, Windows-1252, Macintosh). |
| `--name-transform decoder=ENCODING` | Decodes the file name from the specified encoding. |
| `--name-transform charmap=MAP` | Applies a character mapping transformation. |
| `--name-transform lowercase` | Converts the file name to lowercase. |
| `--name-transform uppercase` | Converts the file name to UPPERCASE. |
| `--name-transform titlecase` | Converts the file name to Title Case. |
| `--name-transform ascii` | Strips non-ASCII characters. |
| `--name-transform url` | URL-encodes the file name. |
| `--name-transform nfc` | Converts the file name to NFC Unicode normalization form. |
| `--name-transform nfd` | Converts the file name to NFD Unicode normalization form. |
| `--name-transform nfkc` | Converts the file name to NFKC Unicode normalization form. |
| `--name-transform nfkd` | Converts the file name to NFKD Unicode normalization form. |
| `--name-transform command=/path/to/my/programfile names.` | Executes an external program to transform |


Conversion modes:  
```
none  
nfc  
nfd  
nfkc  
nfkd  
replace  
prefix  
suffix  
suffix_keep_extension  
trimprefix  
trimsuffix  
index  
date  
truncate  
base64encode  
base64decode  
encoder  
decoder  
ISO-8859-1  
Windows-1252  
Macintosh  
charmap  
lowercase  
uppercase  
titlecase  
ascii  
url  
regex  
command  
```
Char maps:  
```
  
IBM-Code-Page-037  
IBM-Code-Page-437  
IBM-Code-Page-850  
IBM-Code-Page-852  
IBM-Code-Page-855  
Windows-Code-Page-858  
IBM-Code-Page-860  
IBM-Code-Page-862  
IBM-Code-Page-863  
IBM-Code-Page-865  
IBM-Code-Page-866  
IBM-Code-Page-1047  
IBM-Code-Page-1140  
ISO-8859-1  
ISO-8859-2  
ISO-8859-3  
ISO-8859-4  
ISO-8859-5  
ISO-8859-6  
ISO-8859-7  
ISO-8859-8  
ISO-8859-9  
ISO-8859-10  
ISO-8859-13  
ISO-8859-14  
ISO-8859-15  
ISO-8859-16  
KOI8-R  
KOI8-U  
Macintosh  
Macintosh-Cyrillic  
Windows-874  
Windows-1250  
Windows-1251  
Windows-1252  
Windows-1253  
Windows-1254  
Windows-1255  
Windows-1256  
Windows-1257  
Windows-1258  
X-User-Defined  
```
Encoding masks:  
```
Asterisk  
 BackQuote  
 BackSlash  
 Colon  
 CrLf  
 Ctl  
 Del  
 Dollar  
 Dot  
 DoubleQuote  
 Exclamation  
 Hash  
 InvalidUtf8  
 LeftCrLfHtVt  
 LeftPeriod  
 LeftSpace  
 LeftTilde  
 LtGt  
 None  
 Percent  
 Pipe  
 Question  
 Raw  
 RightCrLfHtVt  
 RightPeriod  
 RightSpace  
 Semicolon  
 SingleQuote  
 Slash  
 SquareBracket  
```
Examples: 

```
rclone convmv "stories/The Quick Brown Fox!.txt" --name-transform "all,uppercase"
// Output: STORIES/THE QUICK BROWN FOX!.TXT
```

```
rclone convmv "stories/The Quick Brown Fox!.txt" --name-transform "all,replace=Fox:Turtle" --name-transform "all,replace=Quick:Slow"
// Output: stories/The Slow Brown Turtle!.txt
```

```
rclone convmv "stories/The Quick Brown Fox!.txt" --name-transform "all,base64encode"
// Output: c3Rvcmllcw==/VGhlIFF1aWNrIEJyb3duIEZveCEudHh0
```

```
rclone convmv "c3Rvcmllcw==/VGhlIFF1aWNrIEJyb3duIEZveCEudHh0" --name-transform "all,base64decode"
// Output: stories/The Quick Brown Fox!.txt
```

```
rclone convmv "stories/The Quick Brown 🦊 Fox Went to the Café!.txt" --name-transform "all,nfc"
// Output: stories/The Quick Brown 🦊 Fox Went to the Café!.txt
```

```
rclone convmv "stories/The Quick Brown 🦊 Fox Went to the Café!.txt" --name-transform "all,nfd"
// Output: stories/The Quick Brown 🦊 Fox Went to the Café!.txt
```

```
rclone convmv "stories/The Quick Brown 🦊 Fox!.txt" --name-transform "all,ascii"
// Output: stories/The Quick Brown  Fox!.txt
```

```
rclone convmv "stories/The Quick Brown Fox!.txt" --name-transform "all,trimsuffix=.txt"
// Output: stories/The Quick Brown Fox!
```

```
rclone convmv "stories/The Quick Brown Fox!.txt" --name-transform "all,prefix=OLD_"
// Output: OLD_stories/OLD_The Quick Brown Fox!.txt
```

```
rclone convmv "stories/The Quick Brown 🦊 Fox Went to the Café!.txt" --name-transform "all,charmap=ISO-8859-7"
// Output: stories/The Quick Brown _ Fox Went to the Caf_!.txt
```

```
rclone convmv "stories/The Quick Brown Fox: A Memoir [draft].txt" --name-transform "all,encoder=Colon,SquareBracket"
// Output: stories/The Quick Brown Fox： A Memoir ［draft］.txt
```

```
rclone convmv "stories/The Quick Brown 🦊 Fox Went to the Café!.txt" --name-transform "all,truncate=21"
// Output: stories/The Quick Brown 🦊 Fox
```

```
rclone convmv "stories/The Quick Brown Fox!.txt" --name-transform "all,command=echo"
// Output: stories/The Quick Brown Fox!.txt
```

```
rclone convmv "stories/The Quick Brown Fox!" --name-transform "date=-{YYYYMMDD}"
// Output: stories/The Quick Brown Fox!-20250504
```

```
rclone convmv "stories/The Quick Brown Fox!" --name-transform "date=-{macfriendlytime}"
// Output: stories/The Quick Brown Fox!-2025-05-04 0448AM
```

```
rclone convmv "stories/The Quick Brown Fox!.txt" --name-transform "all,regex=[\\.\\w]/ab"
// Output: ababababababab/ababab ababababab ababababab ababab!abababab
```



Multiple transformations can be used in sequence, applied in the order they are specified on the command line.

The `--name-transform` flag is also available in `sync`, `copy`, and `move`.

# Files vs Directories ##

By default `--name-transform` will only apply to file names. The means only the leaf file name will be transformed.
However some of the transforms would be better applied to the whole path or just directories.
To choose which which part of the file path is affected some tags can be added to the `--name-transform`

| Tag | Effect |
|------|------|
| `file` | Only transform the leaf name of files (DEFAULT) |
| `dir` | Only transform name of directories - these may appear anywhere in the path |
| `all` | Transform the entire path for files and directories |

This is used by adding the tag into the transform name like this: `--name-transform file,prefix=ABC` or `--name-transform dir,prefix=DEF`.

For some conversions using all is more likely to be useful, for example `--name-transform all,nfc`

Note that `--name-transform` may not add path separators `/` to the name. This will cause an error.

# Ordering and Conflicts ##

* Transformations will be applied in the order specified by the user.
    * If the `file` tag is in use (the default) then only the leaf name of files will be transformed.
    * If the `dir` tag is in use then directories anywhere in the path will be transformed
    * If the `all` tag is in use then directories and files anywhere in the path will be transformed
    * Each transformation will be run one path segment at a time.
    * If a transformation adds a `/` or ends up with an empty path segment then that will be an error.
* It is up to the user to put the transformations in a sensible order.
    * Conflicting transformations, such as `prefix` followed by `trimprefix` or `nfc` followed by `nfd`, are possible.
    * Instead of enforcing mutual exclusivity, transformations are applied in sequence as specified by the
user, allowing for intentional use cases (e.g., trimming one prefix before adding another).
    * Users should be aware that certain combinations may lead to unexpected results and should verify
transformations using `--dry-run` before execution.

# Race Conditions and Non-Deterministic Behavior ##

Some transformations, such as `replace=old:new`, may introduce conflicts where multiple source files map to the same destination name.
This can lead to race conditions when performing concurrent transfers. It is up to the user to anticipate these.
* If two files from the source are transformed into the same name at the destination, the final state may be non-deterministic.
* Running rclone check after a sync using such transformations may erroneously report missing or differing files due to overwritten results.

* To minimize risks, users should:
    * Carefully review transformations that may introduce conflicts.
    * Use `--dry-run` to inspect changes before executing a sync (but keep in mind that it won't show the effect of non-deterministic transformations).
    * Avoid transformations that cause multiple distinct source files to map to the same destination name.
    * Consider disabling concurrency with `--transfers=1` if necessary.
    * Certain transformations (e.g. `prefix`) will have a multiplying effect every time they are used. Avoid these when using `bisync`.

	

```
rclone convmv dest:path --name-transform XXX [flags]
```

ncw commented:
Great job @nielash :-)

I pushed a doc fix onto branch [pr-8538-transform](https://github.com/rclone/rclone/tree/pr-8538-transform) (I couldn't update your branch due to "repository rule violations" whatever that is! This has a commit from master and a commit for this branch. I realised when reviewing it that I made a mistake with `max_connections` and you copied the mistake for `name_transform` - we just one one line docs in `fs/config.go` and the full docs go in `docs/content/docs.md`.

I found a couple of oddities - otherwise it seems to work well :-)

----

First - odd debugs with 🦊

```
$ rclone copy -vv --name-transform sausage /tmp/file.txt  /tmp/
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!.txt: transformed to: STORIES/THE QUICK BROWN FOX!.TXT
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!.txt: transformed to: stories/The Slow Brown Turtle!.txt
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!.txt: transformed to: c3Rvcmllcw==/VGhlIFF1aWNrIEJyb3duIEZveCEudHh0
2025/05/06 14:31:25 DEBUG : c3Rvcmllcw==/VGhlIFF1aWNrIEJyb3duIEZveCEudHh0: transformed to: stories/The Quick Brown Fox!.txt
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown 🦊 Fox Went to the Café!.txt: transformed to: stories/The Quick Brown 🦊 Fox Went to the Café!.txt
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown 🦊 Fox!.txt: transformed to: stories/The Quick Brown  Fox!.txt
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!.txt: transformed to: stories/The Quick Brown Fox!
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!.txt: transformed to: OLD_stories/OLD_The Quick Brown Fox!.txt
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown 🦊 Fox Went to the Café!.txt: transformed to: stories/The Quick Brown _ Fox Went to the Caf_!.txt
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox: A Memoir [draft].txt: transformed to: stories/The Quick Brown Fox： A Memoir ［draft］.txt
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown 🦊 Fox Went to the Café!.txt: transformed to: stories/The Quick Brown 🦊 Fox
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!: transformed to: stories/The Quick Brown Fox!-20250506
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!: transformed to: stories/The Quick Brown Fox!-2025-05-06 0231PM
2025/05/06 14:31:25 DEBUG : stories/The Quick Brown Fox!.txt: transformed to: ababababababab/ababab ababababab ababababab ababab!abababab
2025/05/06 14:31:25 DEBUG : rclone: Version "v1.70.0-beta.8730.53bdd5808.pr-8538-transform" starting with parameters ["rclone" "copy" "-vv" "--name-transform" "sausage" "/tmp/file.txt" "/tmp/"]
```

----

Next - this error handling is strange. I guess that is a consequence of lib/transform having nowhere to report errors? It ends up copying the file without doing the name transform. Maybe we need to re-visit the error handling :-( It could use `fs.CountError` with a non retriable error which would at least stop the sync being retried, but I think it has copied the file before then hasn't it.

```
$ rclone copy -vv --name-transform all,sausage /tmp/file.txt  /tmp/dst/
[snip]
2025/05/06 14:32:27 ERROR : file.txt: Failed to parse transform flags: invalid choice "sausage" from: none, nfc, nfd, nfkc, nfkd, replace, prefix, suffix, suffix_keep_extension, trimprefix, trimsuffix, index, date, truncate, base64encode, base64decode, encoder, decoder, ISO-8859-1, Windows-1252, Macintosh, charmap, lowercase, uppercase, titlecase, ascii, url, regex, command
2025/05/06 14:32:27 DEBUG : file.txt: md5 = b1946ac92492d2347c6235b4d2611184 OK
2025/05/06 14:32:27 DEBUG : file.txt.ae51307.partial: renamed to: file.txt
2025/05/06 14:32:27 INFO  : file.txt: Copied (new)
2025/05/06 14:32:27 ERROR : Attempt 1/3 failed with 1 errors and: invalid choice "sausage" from: none, nfc, nfd, nfkc, nfkd, replace, prefix, suffix, suffix_keep_extension, trimprefix, trimsuffix, index, date, truncate, base64encode, base64decode, encoder, decoder, ISO-8859-1, Windows-1252, Macintosh, charmap, lowercase, uppercase, titlecase, ascii, url, regex, command
2025/05/06 14:32:27 DEBUG : file.txt: Size and modification time the same (differ by 0s, within tolerance 1ns)
2025/05/06 14:32:27 DEBUG : file.txt: Unchanged skipping
2025/05/06 14:32:27 ERROR : Attempt 2/3 succeeded
2025/05/06 14:32:27 INFO  :
Transferred:              6 B / 6 B, 100%, 0 B/s, ETA -
Transferred:            1 / 1, 100%
Elapsed time:         0.0s

2025/05/06 14:32:27 DEBUG : 6 go routines active
```

----

nielash commented:
> First - odd debugs with 🦊

Haha! That must be from generating the help text for the `convmv` command on `init`. I can quiet the logs easily enough, but it would be even better to just prevent it going down that codepath unless the help text is actually needed... is there an easy way to detect that with cobra? I can experiment... 

Could also make it static text of course...but that's not as fun.

> Next - this error handling is strange. I guess that is a consequence of lib/transform having nowhere to report errors? It ends up copying the file without doing the name transform. Maybe we need to re-visit the error handling :-( It could use `fs.CountError` with a non retriable error which would at least stop the sync being retried, but I think it has copied the file before then hasn't it.

Hmm... yeah that's not great. What is tough is that there are places like [here](https://github.com/nielash/rclone/blob/53bdd5808575033782a086e5e54576d7428c7409/fs/march/march.go#L90) where it would be hard to use the error even if we returned it. Would canceling the context be useful?

nielash commented:
> I don't think cancelling the context will help (in fact you can't) but doing a non retriable error will help.
> 
> An ERROR log + `fs.CountError` + Non retriable error should be enough.
> 
> If all else fails then we'll have to plumb in error handling, but let's not do that just yet!

I implemented that here: https://github.com/rclone/rclone/pull/8538/commits/0eb462fd46cb51996f1ee376b3228b760fffda49 (sorry, probably should have commented on this thread to mention that!)



> I don't know. I think all the help text needs to be generated in advance, but cobra is quite complicated so there may be a way to defer it.

Added a quick-fix here: https://github.com/rclone/rclone/pull/8538/commits/fd05dc116c0dbacb8730e225ab41e62351f57061 although still strikes me as clunky...

ncw commented:
Thanks for doing this @nielash 

I've tidied the commits up a bit and merged it to master now.

I couldn't push it back to your branch (some permissions problem) to update this PR so I've merged it directly.
